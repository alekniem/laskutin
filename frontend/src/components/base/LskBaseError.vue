<template>
    <div class="red--text font-weight-bold">
        <slot />
    </div>
</template>

<script>
export default {
    name: "LskBaseError",
};
</script>
