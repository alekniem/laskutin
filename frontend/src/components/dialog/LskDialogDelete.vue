<template>
    <v-dialog max-width="700" persistent v-model="dialogVisible">
        <ValidationObserver #default="{ handleSubmit, invalid }" ref="observer">
            <v-form>
                <v-card>
                    <v-card-title :id="cid + 'Title'">
                        {{ title | trans }} ({{ id }})
                    </v-card-title>
                    <v-card-text>
                        <LskInputCheckbox
                            :cid="cid + 'YesCheckboxInput'"
                            :label="trans(checkboxLabel)"
                            vid="yes_checkbox"
                            v-model="checkbox"
                        />
                        <LskBaseError v-if="apiError">
                            {{ apiError | trans }}
                        </LskBaseError>
                    </v-card-text>
                    <v-card-actions>
                        <v-spacer />
                        <LskInputButton
                            @click.native="processCancel"
                            :cid="cid + 'CancelButton'"
                            buttonText="Cancel"
                        />
                        <LskInputButton
                            @click.native="handleSubmit(processDelete)"
                            :cid="cid + 'DeleteButton'"
                            :disabled="invalid || deleteButtonDisabled"
                            buttonText="Delete"
                        />
                    </v-card-actions>
                </v-card>
            </v-form>
        </ValidationObserver>
    </v-dialog>
</template>

<script>
import lsk_api from "@/plugins/lsk_api";
import { ValidationObserver } from "vee-validate";

export default {
    name: "LskDialogDelete",

    components: {
        ValidationObserver,
    },

    model: {
        prop: "showDialog",
        event: "dialog",
    },

    props: {
        apiActionName: { type: String, required: true },
        checkboxLabel: { type: String, required: true },
        cid: { type: String, required: true },
        id: { type: [String, Number], required: false },
        showDialog: { type: Boolean, required: true },
        title: { type: String, required: true },
    },

    data() {
        return {
            apiError: null,
            checkbox: false,
            dialogVisible: this.showDialog,
        };
    },

    computed: {
        deleteButtonDisabled() {
            return !this.checkbox;
        },
    },

    watch: {
        showDialog(value) {
            this.dialogVisible = value;
        },
    },

    methods: {
        processCancel() {
            this.$emit("dialog", false);
        },

        processDelete() {
            lsk_api
                .query(this.apiActionName, this.id, null)
                .then((response) => this.parseResponse(response.data))
                .catch((error) => this.parseError(error.response.data));
        },

        parseResponse() {
            this.$emit("dialog", false);
        },

        parseError(data) {
            this.apiError = data.message || "Failed to delete item";
        },
    },
};
</script>
