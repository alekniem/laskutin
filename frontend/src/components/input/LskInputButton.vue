<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
