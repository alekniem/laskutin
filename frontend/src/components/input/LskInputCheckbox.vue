<template>
    <ValidationProvider
        :customMessages="{ required: trans('Required') }"
        :rules="{ required: { allowFalse: false } }"
        :vid="vid"
        ref="provider"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-checkbox
            @change="$emit('change', $event)"
            :error-messages="errors"
            :id="cid"
            :input-value="checked"
            :label="trans(label)"
            :success="valid"
        />
    </ValidationProvider>
</template>

<script>
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputCheckbox",

    components: {
        ValidationProvider,
    },

    model: {
        prop: "checked",
        event: "change",
    },

    props: {
        cid: { type: String, required: true },
        checked: { type: Boolean, required: true },
        label: { type: String, required: true },
        vid: { type: String, required: true },
    },
};
</script>
