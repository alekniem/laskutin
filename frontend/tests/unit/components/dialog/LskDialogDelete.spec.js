import axios from "axios";
import flushPromises from "flush-promises";
import LskBaseError from "@/components/base/LskBaseError";
import LskDialogDelete from "@/components/dialog/LskDialogDelete";
import LskInputButton from "@/components/input/LskInputButton";
import LskInputCheckbox from "@/components/input/LskInputCheckbox";
import { createWrapper } from "UT/lsk_test_helper";

jest.mock("axios");

axios.delete.mockImplementation((url) => {
    if (url === "customer/123") {
        return Promise.resolve({
            data: {
                message: "DELETE QUERY OK",
            },
        });
    }
    return Promise.reject({
        response: {
            data: {
                message: "DELETE QUERY ERROR",
            },
        },
    });
});

const getPropsData = ({
    apiActionName = "customerDestroy",
    checkboxLabel = "TEST CHECKBOX LABEL",
    cid = "TEST_CID",
    id = 123,
    showDialog = true,
    title = "TEST TITLE",
} = {}) => ({
    apiActionName,
    checkboxLabel,
    cid,
    id,
    showDialog,
    title,
});

const stubs = {
    LskBaseError,
    LskInputButton,
    LskInputCheckbox,
};

describe("LskDialogDelete", () => {
    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData({ showDialog: false }),
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog not visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData(),
            stubs,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData(),
            stubs,
        });
        await flushPromises();
        wrapper.vm._data.apiError = "API ERROR";
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("error messages");
    });

    it("should change 'delete' button from disabled to enabled", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData(),
            stubs,
        });
        await flushPromises();

        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.find("#TEST_CIDDeleteButton").attributes().disabled).toBeDefined();

        wrapper.find("#TEST_CIDYesCheckboxInput").trigger("click");
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.find("#TEST_CIDDeleteButton").attributes().disabled).toBeUndefined();
    });

    it("should emit 'dialog' event on success", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData(),
            stubs,
        });
        await flushPromises();
        wrapper.find("#TEST_CIDYesCheckboxInput").trigger("click");
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        wrapper.find("#TEST_CIDDeleteButton").trigger("click");
        await flushPromises();
        expect(wrapper.emitted().dialog).toStrictEqual([[false]]);
    });

    it("should show an error message on failure", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData({ id: 456 }),
            stubs,
        });
        await flushPromises();
        wrapper.find("#TEST_CIDYesCheckboxInput").trigger("click");
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        wrapper.find("#TEST_CIDDeleteButton").trigger("click");
        await flushPromises();
        expect(wrapper.vm._data.apiError).toStrictEqual("DELETE QUERY ERROR");
    });

    it("should emit 'dialog' event on cancel", async () => {
        const wrapper = createWrapper(LskDialogDelete, {
            propsData: getPropsData(),
            stubs,
        });
        await flushPromises();
        wrapper.find("#TEST_CIDCancelButton").trigger("click");
        await flushPromises();
        expect(wrapper.emitted().dialog).toStrictEqual([[false]]);
    });
});
