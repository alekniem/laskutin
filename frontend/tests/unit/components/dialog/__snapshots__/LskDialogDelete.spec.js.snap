// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskDialogDelete should match snapshot: dialog not visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
</div>
`;

exports[`LskDialogDelete should match snapshot: dialog visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                TEST TITLE (123)
              </div>
              <div class="v-card__text">
                <div>
                  <div class="v-input theme--light v-input--selection-controls v-input--checkbox">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-input--selection-controls__input">
                          <i aria-hidden="true"
                             class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"
                          >
                          </i>
                          <input aria-checked="false"
                                 id="TEST_CIDYesCheckboxInput"
                                 role="checkbox"
                                 type="checkbox"
                                 value
                          >
                          <div class="v-input--selection-controls__ripple">
                          </div>
                        </div>
                        <label for="TEST_CIDYesCheckboxInput"
                               class="v-label theme--light"
                               style="left: 0px; position: relative;"
                        >
                          TEST CHECKBOX LABEL
                        </label>
                      </div>
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div"
                                               name="message-transition"
                                               class="v-messages__wrapper"
                        >
                        </transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDDeleteButton"
                >
                  <span class="v-btn__content">
                    Delete
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`LskDialogDelete should match snapshot: error messages 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                TEST TITLE (123)
              </div>
              <div class="v-card__text">
                <div>
                  <div class="v-input v-input--has-state theme--light v-input--selection-controls v-input--checkbox error--text">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-input--selection-controls__input">
                          <i aria-hidden="true"
                             class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light error--text"
                          >
                          </i>
                          <input aria-checked="false"
                                 id="TEST_CIDYesCheckboxInput"
                                 role="checkbox"
                                 type="checkbox"
                                 value
                          >
                          <div class="v-input--selection-controls__ripple error--text">
                          </div>
                        </div>
                        <label for="TEST_CIDYesCheckboxInput"
                               class="v-label theme--light error--text"
                               style="left: 0px; position: relative;"
                        >
                          TEST CHECKBOX LABEL
                        </label>
                      </div>
                      <div class="v-messages theme--light error--text"
                           role="alert"
                      >
                        <transition-group-stub tag="div"
                                               name="message-transition"
                                               class="v-messages__wrapper"
                        >
                          <div class="v-messages__message">
                            Required
                          </div>
                        </transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="red--text font-weight-bold">
                  API ERROR
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDDeleteButton"
                >
                  <span class="v-btn__content">
                    Delete
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;
